tamproxy/comm/packet_forwarder.py

```diff
@@ -58,21 +63,39 @@ def callback_responses(self):
         while self.pipe.poll():
             request, response = self.pipe.recv()
             self.packets_received += 1
-            if response == c.host.reset_msg: 
+            if response == c.host.reset_msg:
                 self.empty_queue()
                 if self.reset_callback: self.reset_callback()
             elif len(response) == 1 and response in c.responses:
                 if response == 'G': continue
-                print ("Firmware responded with an error: {} "
-                       "for the request {}").format(
-                            c.responses[response].msg, request)
+                logger.error(
+                    "Firmware responded with an error, {}, "
+                    "for the request {}".format(
+                        c.responses[response].msg, request
+                    )
+                )
             elif request in self.callback_dict:
                 self.callback_dict[request](request, response)
+            else:
+                logger.warn("Packet recieved with no callback")
 
     def run(self):
         self.pc.start()
-        while not self.__stop.isSet():
+        stopping = False
+        while True:
             self.forward_requests()
             self.callback_responses()
-        # Kill the pc process if thread stopped
-        self.pc.stop()
+
+            # finish any pending packets before stopping
+            if self.__stop.isSet():
+                if not stopping and self.sending_queue.empty():
+                    self.pc.stop()
+                    stopping = True
+
+                elif stopping and not self.pc.is_alive():
```

Comment:
Ah I didn't notice that this case falls through to the loop, my bad. Looks good to me.